src/tools/ceph-dencoder/denc_plugin.h

```diff
@@ -10,7 +10,11 @@ class DencoderPlugin {
   using dencoders_t = std::vector<std::pair<std::string, Dencoder*>>;
 public:
   DencoderPlugin(const fs::path& path) {
+#if defined(__FreeBSD__)
+    mod = dlopen(path.c_str(), RTLD_NOW | RTLD_NODELETE);
```

Comment:
> In contrast to Linux, FreeDSB does not maintain a counter with the
> number of dlopen()'s. So the first dlclose() unloads the lib.
> In Linux only the counter is decremented until it is the last
> dlclose()

all true. but the first call of `dlclose()` is called by the destructor of `DencoderPlugin`. i think that's expected. and if the `dlclose()` call unloads the plugins, that's also fine. why do you want to prevent this from happening?

Reply:
I'm afraid that I do not quite know why the library is actually unloaded before we start calling its code.
Only thing I do know is that this fixes the problem of not being able to call functions.
And then still the closing code get OS warnings like:
```
12066: __cxa_thread_call_dtors: dtr 0x418990 from unloaded dso, skipping
```
Which means that the dl-library is unloaded, there are still calls to its destructors....
I suspect that `_atexit()` plays a role in this, where again dtors are called, but already got unloaded.

Main reason for adding the `RTLD_NODELETE` is that otherwise I get:
```
44058: failed to dlopen("/home/wjw/master/build/lib/denc-mod-rgw.so"): /home/wjw/master/build/lib/denc-mod-rgw.so: invalid file format
```
Perhaps I should add that to the commit?

Reply:
i am not able to connect `RTLD_NODELETE` to `invalid file format`. could you point me to any related document or explain the relation between them?

Reply:
Looking at the man page:
```
     RTLD_LAZY	 Each external function	reference is resolved when the func-
		 tion is first called.

     RTLD_NOW	 All external function references are bound immediately	by
		 dlopen().

     RTLD_LAZY is normally preferred, for reasons of efficiency.  However,
     RTLD_NOW is useful	to ensure that any undefined symbols are discovered
     during the	call to	dlopen().
```
and
```
     RTLD_NODELETE  Prevents unload of the loaded object on dlclose().	The
		    same behaviour may be requested by -z nodelete option of
		    the	static linker ld(1).

```

The error 'invalid file format' stems from:
```
static Elf_Ehdr *
get_elf_header(int fd, const char *path, const struct stat *sbp,
    Elf_Phdr **phdr_p)
{
        Elf_Ehdr *hdr;
        Elf_Phdr *phdr;

        /* Make sure file has enough data for the ELF header */
        if (sbp != NULL && sbp->st_size < (off_t)sizeof(Elf_Ehdr)) {
                _rtld_error("%s: invalid file format", path);
                return (NULL);
        }
...
...
        /* Make sure the file is valid */
        if (!IS_ELF(*hdr)) {
                _rtld_error("%s: invalid file format", path);
                goto error;
        }

```

How that relates to `RTL_NODELETE` is rather unknown to me...
It sort of looks to me that the library is not loaded and that every reference to a dynamic function calls 'dlsym()' for resolution. And one way or another the library was nog kept.
Setting the `RTLD_NODELETE` prevents the unload, and `dlsym()` is able to resolve.

I guess that to get to the bottom of this, there is a lot of low level libc debugging required.
Setting `RTLD_NODELETE` is a sort of poor mans emulation of the Linux behaviour. The difference being that the library is never unloaded.

Reply:
these are facts. none of them is directly related to your fix. and your commit message claims "fixes: #42607", so presumably #42607 introduced a regression, am i right? does ceph-dencoder work on FreeBSD before #42607 ?

Reply:
ceph-dencoder works before #42607 
But I have a set of local patches that augment each `dlopen()` with `RTLD_NODELETE`, since it is still on my long list of things to figure out why this is the problem it is.
So I guess that I perhaps better submit a _work around_ for all that I have in my local patches. And not as a fix on #42607

Reply:
it's confusing. by "it worked before #42607", i assume you meant to say it worked with the `RTLD_NODELETE` patches applied. but what you are proposing is to apply a "`RTLD_NODELETE`" change. so i am not sure #42607 introduced a regression. IIUC, it's just an issue which existed **before** #42607.

Reply:
I agree that it is not a regression, but something I should have submitted a PR for some time ago.
And it does need my patches to achtually work.

Reply:
Tracing the syscalls code of ceph-dencoder I get:
```
open("/home/wjw/master/build/lib/denc-mod-common.so",O_RDONLY|O_CLOEXEC|O_VERIFY,016213266556) = 4 (0x4)
fstat(4,{ mode=-rwxrwxr-x ,inode=2509565,size=36870064,blksize=131072 }) = 0 (0x0)
fstatfs(4,{ fstypename=zfs,mntonname=/,mntfromname=zroot/ROOT/default,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,4,0x0) = 34366464000 (0x80066a000)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE,4,0x450000)  = 34366468096 (0x80066b000)
mmap(0x0,6705152,PROT_NONE,MAP_GUARD|MAP_ALIGNED_SUPER,-1,0x0) = 34426847232 (0x804000000)
mmap(0x804000000,4526080,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,4,0x0) = 34426847232 (0x8040000
00)
mmap(0x804651000,65536,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,4,0x451000) = 34433470464 (0x804651000)
mmap(0x804661000,16384,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_ANON,-1,0x0) = 34433536000 (0x804661000)
munmap(0x80066b000,4096)                         = 0 (0x0)
munmap(0x80066a000,4096)                         = 0 (0x0)
close(4)                                         = 0 (0x0)
getdirentries(3,"k;&\0\0\0\0\0\^Z\M^^\M-)\^V\0\0"...,4096,{ 0x16a649f6 }) = 4088 (0xff8)
open("/home/wjw/master/build/lib/denc-mod-rbd.so",O_RDONLY|O_CLOEXEC|O_VERIFY,037677667162) = 4 (0x4)
fstat(4,{ mode=-rwxrwxr-x ,inode=2507900,size=19893216,blksize=131072 }) = 0 (0x0)
fstatfs(4,{ fstypename=zfs,mntonname=/,mntfromname=zroot/ROOT/default,fsid= }) = 0 (0x0)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_PREFAULT_READ,4,0x0) = 34366464000 (0x80066a000)
mmap(0x0,4096,PROT_READ,MAP_PRIVATE,4,0x33d000)  = 34366468096 (0x80066b000)
mmap(0x0,5550080,PROT_NONE,MAP_GUARD|MAP_ALIGNED_SUPER,-1,0x0) = 34435235840 (0x804800000)
mmap(0x804800000,3399680,PROT_READ|PROT_EXEC,MAP_PRIVATE|MAP_FIXED|MAP_NOCORE|MAP_PREFAULT_READ,4,0x0) = 34435235840 (0x8048000
00)
mmap(0x804d3d000,57344,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_PREFAULT_READ,4,0x33d000) = 34440728576 (0x804d3d000)
munmap(0x80066b000,4096)                         = 0 (0x0)
munmap(0x80066a000,4096)                         = 0 (0x0)
close(4)     
```

Which repeats itself for all library loads. Which seems to confirm my expectations.
It does not answer the why?
I'll have to ask more knowledgable people in the FreeBSD community.